### ISSUE TITLE
[apex] False positive ApexSharingViolationsRule, unsupported Apex feature

### ISSUE THREAD

**Issue by nchursin** [pmd/pmd#1664]
**Affects PMD Version:** PMD 6.11.0

**Rule:** [ApexSharingViolationsRule](https://pmd.github.io/pmd-6.11.0/pmd_rules_apex_security.html#apexsharingviolations)

**Description:**
The rule doesn't support the new `inherited sharing` keyword. This is because currently PMD uses `OldModifiers` lib for checking modifiers, but it's not going to be updated with the new features (according to its name). The `inherited sharing` keyword was introduced in Winter '19 release.

**Code Sample demonstrating the issue:**

```
public inherited sharing class Aq {
}
```

**Running PMD through:** Ant, VSCode plugin

**Comment by adangel** [comment]
Thanks for reporting!
It appears, that there is a new jorje library available: https://github.com/forcedotcom/salesforcedx-vscode/tree/develop/packages/salesforcedx-vscode-apex/out
We last updated jorje on 2019-01-18.
Unfortunately that's a blob and I have no clue, what has been changed there....

The documentation section about `inherited sharing` indeed appears only for Winter '19 release (v44) and later.

**Comment by jsotuyod** [comment]
The problem lies here:

https://github.com/pmd/pmd/blob/74cf7dc77b2c26e8a4bb0660690c3df871bead93/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexSharingViolationsRule.java#L104-L114

The check is correct for `sharing` and `without sharing`, but the newly added `inherited sharing` is not listed under the `OldModifiers.ModifierType` constants.

I've found nowhere in the Apex Jorje lib a constant we can use for this new case. It seems we will have to hardwire a `"inheritedSharing"` check along the existing ones.

**Comment by adangel** [comment]
Instead of `OldModifiers.ModifierType` we should use `apex.jorje.semantic.symbol.type.ModifierTypeInfos` - this contains constants for "WITH_SHARING", "WITHOUT_SHARING" as well as "INHERITED_SHARING".

I've found this by grepping the extracted jorje.jar (`grep -r inheritedSharing *`). "inheritedSharing" is apparently the "bytecodeName".

**Comment by adangel** [comment]
With the changes from #1727 the fix is now trivial. It is about this line:

https://github.com/pmd/pmd/blob/bb071d68e24ac1cad784af18b60ee1af0effb9c9/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexSharingViolationsRule.java#L99

Just adding another condition (`node.getModifiers().isInheritedSharing()`) should fix this issue.
Of course, a new test case needs to be added as well.

**Comment by cnaccio** [comment]
Hey guys, any updates on this? @adangel Is there a workaround we can use without having to recompile pmd?

**Comment by adangel** [comment]
I don't think so, there is a code fix required. This issue is up for grabs (labeled good first issue). Any volunteers?

**Comment by rsoesemann** [comment]
@nchursin we just fixed this and it will be available in the release that is coming in the next days. Thanks for reporting that important issue.

**Comment by nchursin** [comment]
Works as charm, thank you!  Just got time to test it 🙂
